filebeat/input/kafka/input.go

```diff
@@ -21,96 +21,161 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
+	"io"
 	"strings"
 	"sync"
 	"time"
 
+	"github.com/elastic/beats/v7/libbeat/common/atomic"
+
 	"github.com/Shopify/sarama"
+	"github.com/pkg/errors"
 
-	"github.com/elastic/beats/v7/filebeat/channel"
-	"github.com/elastic/beats/v7/filebeat/input"
+	input "github.com/elastic/beats/v7/filebeat/input/v2"
 	"github.com/elastic/beats/v7/libbeat/beat"
 	"github.com/elastic/beats/v7/libbeat/common"
 	"github.com/elastic/beats/v7/libbeat/common/acker"
 	"github.com/elastic/beats/v7/libbeat/common/backoff"
 	"github.com/elastic/beats/v7/libbeat/common/kafka"
+	"github.com/elastic/beats/v7/libbeat/feature"
 	"github.com/elastic/beats/v7/libbeat/logp"
-
-	"github.com/pkg/errors"
+	"github.com/elastic/beats/v7/libbeat/reader"
+	"github.com/elastic/beats/v7/libbeat/reader/parser"
 )
 
-func init() {
-	err := input.Register("kafka", NewInput)
+const pluginName = "kafka"
+
+// Plugin creates a new filestream input plugin for creating a stateful input.
+func Plugin() input.Plugin {
+	return input.Plugin{
+		Name:       pluginName,
+		Stability:  feature.Stable,
+		Deprecated: false,
+		Info:       "Kafka input",
+		Doc:        "The Kafka input consumes events from topics by connecting to the configured kafka brokers",
+		Manager:    input.ConfigureWith(configure),
+	}
+}
+
+func configure(cfg *common.Config) (input.Input, error) {
+	config := defaultConfig()
+	if err := cfg.Unpack(&config); err != nil {
+		return nil, err
+	}
+
+	saramaConfig, err := newSaramaConfig(config)
 	if err != nil {
-		panic(err)
+		return nil, errors.Wrap(err, "initializing Sarama config")
 	}
+	return NewInput(config, saramaConfig)
+}
+
+func NewInput(config kafkaInputConfig, saramaConfig *sarama.Config) (*kafkaInput, error) {
+	return &kafkaInput{config: config, saramaConfig: saramaConfig}, nil
 }
 
-// Input contains the input and its config
 type kafkaInput struct {
 	config          kafkaInputConfig
 	saramaConfig    *sarama.Config
-	context         input.Context
-	outlet          channel.Outleter
 	saramaWaitGroup sync.WaitGroup // indicates a sarama consumer group is active
-	log             *logp.Logger
-	runOnce         sync.Once
 }
 
-// NewInput creates a new kafka input
-func NewInput(
-	cfg *common.Config,
-	connector channel.Connector,
-	inputContext input.Context,
-) (input.Input, error) {
+func (input *kafkaInput) Name() string { return pluginName }
 
-	config := defaultConfig()
-	if err := cfg.Unpack(&config); err != nil {
-		return nil, errors.Wrap(err, "reading kafka input config")
-	}
+func (input *kafkaInput) Test(_ input.TestContext) error {
```

Comment:
Is there a way maybe we could test connecting to Kafka?

Reply:
Added a check that retrieves all topics and checks it against the configured topics.